src/main/scala/org/broadinstitute/dsde/workbench/sam/api/ManagedGroupRoutes.scala

```diff
@@ -1,59 +1,131 @@
+
+      pathEndOrSingleSlash {
+        get {
+          handleGetGroup(managedGroup)
+        } ~
+        post {
+          handlePostGroup(managedGroup, userInfo)
+        } ~
+        delete {
+          handleDeleteGroup(managedGroup, userInfo)
+        }
       } ~
-      delete {
-        handleDeleteGroup(groupId, userInfo)
+      pathPrefix(Segment) { policyName =>
+        val accessPolicyName = parsePolicyName(policyName)
+
+        pathEndOrSingleSlash {
+          get {
+            handleListEmails(managedGroup, accessPolicyName, userInfo)
+          } ~
+          put {
+            handleOverwriteEmails(managedGroup, accessPolicyName, userInfo)
+          }
+        } ~
+        pathPrefix(Segment) { email =>
+          pathEndOrSingleSlash {
+            put {
+              handleAddEmailToPolicy(managedGroup, accessPolicyName, email, userInfo)
+            } ~
+            delete {
+              handleDeleteEmailFromPolicy(managedGroup, accessPolicyName, email, userInfo)
+            }
+          }
+        }
       }
     }
   }
 
-  private def handleGetGroup(groupId: String) = {
+  private def parsePolicyName(policyName: String): ManagedGroupPolicyName = {
+    if (policyName == "members") {
+      ManagedGroupService.memberPolicyName
+    } else if (policyName == "admins") {
+      ManagedGroupService.adminPolicyName
+    } else {
+      throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, "Policy name for managed groups must be one of: [\"admins\", \"members\"]"))
+    }
+  }
+
+  private def handleGetGroup(managedGroup: Resource): Route = {
     complete (
-      managedGroupService.loadManagedGroup(ResourceId(groupId)).map {
+      managedGroupService.loadManagedGroup(managedGroup.resourceId).map {
         case Some(response) => StatusCodes.OK -> response
         case None => throw new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.NotFound, "group not found"))
       }
     )
   }
 
-  private def handlePostGroup(groupId: String, userInfo: UserInfo) = {
-    complete(managedGroupService.createManagedGroup(ResourceId(groupId), userInfo).map(_ => StatusCodes.Created))
+  private def handlePostGroup(managedGroup: Resource, userInfo: UserInfo): Route = {
+    complete(managedGroupService.createManagedGroup(managedGroup.resourceId, userInfo).map(_ => StatusCodes.Created))
+  }
+
+  private def handleDeleteGroup(managedGroup: Resource, userInfo: UserInfo): Route = {
+    requireAction(managedGroup, SamResourceActions.delete, userInfo) {
+      complete(managedGroupService.deleteManagedGroup(managedGroup.resourceId).map(_ => StatusCodes.NoContent))
+    }
+  }
+
+  private def handleListEmails(managedGroup: Resource, accessPolicyName: ManagedGroupPolicyName, userInfo: UserInfo): Route = {
+    requireAction(managedGroup, SamResourceActions.readPolicy(accessPolicyName.asInstanceOf[AccessPolicyName]), userInfo) {
```

Comment:
I don't like the asInstanceOf, why do you need to cast?

you probably want the signature to be `AccessPolicyName with ManagedGroupPolicyName`

and you can tighten that up by declaring a type

Reply:
fixed